client/client_test.go

```diff
@@ -1370,64 +1320,117 @@ func testPublishNoOneHasSnapshotKey(t *testing.T, rootType string) {
-	testPublishBadExistingSnapshot(t, data.ECDSAKey, true, true)
-	testPublishBadExistingSnapshot(t, data.ECDSAKey, false, true)
-}
+	ts := fullTestServer(t)
+	defer ts.Close()
 
-// If the snapshot metadata is unreadable, whether the client or server has the
-// snapshot key, we can't publish.
-// We test this with both an RSA and ECDSA root key
-func TestPublishSnapshotUnreadable(t *testing.T) {
-	testPublishBadExistingSnapshot(t, data.ECDSAKey, true, false)
-	testPublishBadExistingSnapshot(t, data.ECDSAKey, false, false)
+	// do not publish first - publish should fail
+	repo, _ := initializeRepo(t, data.ECDSAKey, "docker.com/notary1", ts.URL, true)
+	defer os.RemoveAll(repo.baseDir)
+	testPublishBadMetadata(t, data.CanonicalSnapshotRole, repo, false, false)
+
+	repo, _ = initializeRepo(t, data.ECDSAKey, "docker.com/notary2", ts.URL, false)
+	defer os.RemoveAll(repo.baseDir)
+	testPublishBadMetadata(t, data.CanonicalSnapshotRole, repo, false, false)
+
+	// publish first - should succeed
+	repo, _ = initializeRepo(t, data.ECDSAKey, "docker.com/notary3", ts.URL, true)
+	defer os.RemoveAll(repo.baseDir)
+	testPublishBadMetadata(t, data.CanonicalSnapshotRole, repo, true, true)
+
+	repo, _ = initializeRepo(t, data.ECDSAKey, "docker.com/notary4", ts.URL, false)
+	defer os.RemoveAll(repo.baseDir)
+	testPublishBadMetadata(t, data.CanonicalSnapshotRole, repo, true, true)
 }
 
-func testPublishBadExistingSnapshot(t *testing.T, rootType string,
-	serverManagesSnapshot bool, readable bool) {
+// If the targets metadata is corrupt or the targets metadata is unreadable,
+// we can't publish for the first time, because there is no existing data for.
+// us to download. If the repo has already been published, it doesn't matter
+// if the metadata is corrupt because we can just redownload if it is.
+func TestPublishTargetsCorrupt(t *testing.T) {
+	ts := fullTestServer(t)
+	defer ts.Close()
+
+	// do not publish first - publish should fail
+	repo, _ := initializeRepo(t, data.ECDSAKey, "docker.com/notary1", ts.URL, false)
+	defer os.RemoveAll(repo.baseDir)
+	testPublishBadMetadata(t, data.CanonicalTargetsRole, repo, false, false)
 
-	// Temporary directory where test files will be created
-	tempBaseDir, err := ioutil.TempDir("/tmp", "notary-test-")
-	defer os.RemoveAll(tempBaseDir)
-	assert.NoError(t, err, "failed to create a temporary directory: %s", err)
+	// publish first - should succeed
+	repo, _ = initializeRepo(t, data.ECDSAKey, "docker.com/notary2", ts.URL, false)
+	defer os.RemoveAll(repo.baseDir)
+	testPublishBadMetadata(t, data.CanonicalTargetsRole, repo, true, true)
+}
+
+// If the root metadata is corrupt or the root metadata is unreadable,
+// we can't publish for the first time or the second time.  Root is the most
+// important and what we used to pin trust, so if it's corrupt, we can't
+// verify downloaded updates.
+func TestPublishRootCorrupt(t *testing.T) {
+	t.Skip("Test currently fails - not sure what the correct behavior is.")
```

Comment:
You're right, we would lose the protection against rollbacks, but since it's a local corruption, I think that behavior might be ok.